generator/src/main/java/com/graphicsfuzz/generator/mutateapi/ExpressionGenerator.java

```diff
@@ -0,0 +1,632 @@
+        + freshId();
+  }
+
+  private String genParamName(Type type, boolean unknownValue) {
+    // Provides a parameter name required when making a paramater of a newly generated function.
+    // For example:
+    //  _GLF_PARAM_UNKNOWN_VALUE_vec4_id_1: a parameter of unknown value of vec4 type.
+    //  _GLF_PARAM_int_id_62: a parameter of integer type.
+    return (unknownValue ? Constants.GLF_PARAM_UNKNOWN_VALUE : Constants.GLF_PARAM)
+        + "_" + type.toString()
+        + freshId();
+  }
+
+  /**
+   * Utility function to parse name from the value given, if value is known, for example, -0.45
+   * will be parsed as negative_0_45. If value is unknown, return the "UNKNOWN VALUE" string.
+   *
+   * @param value value that will be converted to the name of variables or functions.
+   * @return a string derived from the given value which will be used as function or variable names.
+   */
+  private String parseNameFromValue(Value value) {
+    if (value.valueIsUnknown()) {
+      return "_" + value.toString();
+    }
+    final StringBuilder name = new StringBuilder();
+    if (value instanceof NumericValue) {
+      final NumericValue numericValue = (NumericValue) value;
+      float floatValue = numericValue.getValue().get().floatValue();
+      if (floatValue < 0.0) {
+        name.append(NEGATIVE);
+        floatValue = Math.abs(floatValue);
+      }
+      name.append("_");
+      // Replace dot with underscore, i.e., 0.45 will be converted to 0_45.
+      name.append(Float.toString(floatValue).replace(".", "_"));
+
+    }
+    if (value instanceof BooleanValue) {
+      name.append("_").append(value.toString());
+    }
+    return name.toString();
+  }
+
+  private Expr generateVariableFact(FactManager factManager,
+                                    FunctionDefinition currentFunction,
+                                    Stmt stmt,
+                                    Value value) {
+
+    boolean atGlobalScope = factManager.globalScope() || generator.nextBoolean();
+    final String varName = genVarName(value, atGlobalScope);
+
+    final VariableDeclInfo variableDeclInfo = new VariableDeclInfo(varName, null,
+        new Initializer(generateExpr(atGlobalScope ? globalFactManager : factManager,
+            currentFunction,
+            stmt,
+            value
+        )));
+
+    final VariablesDeclaration variablesDecl = new VariablesDeclaration(value.getType(),
+        variableDeclInfo);
+    final VariableDeclFact variableDeclFact = new VariableDeclFact(variablesDecl,
+        variableDeclInfo, value);
+    if (atGlobalScope) {
+      translationUnit.addDeclarationBefore(variablesDecl, currentFunction);
+      globalFactManager.addVariableFact(value, variableDeclFact);
+    } else {
+      currentFunction.getBody().insertBefore(stmt, new DeclarationStmt(variablesDecl));
+      factManager.addVariableFact(value, variableDeclFact);
+    }
+
+
+    return new VariableIdentifierExpr(varName);
+  }
+
```

Comment:
Please add JavaDoc and give 'stmt' a better name.

In general please add JavaDoc for all the 'generate' methods, even though they are not public - they're going to be so important.  And use whatever name you choose to replace 'stmt' with consistently.